### PR TITLE
Output Descriptors

### PR DESCRIPTION
**Pull request by prusnak**
Fixes https://github.com/trezor/trezor-firmware/issues/1362

In order to fully support [Output Descriptors](https://github.com/bitcoin/bitcoin/blob/master/doc/descriptors.md) we need to change things on couple of places:

* **common**
  - [x] add `GetPublicKey.ignore_xpub_magic` and `PublicKey.root_fingerprint` fields
* **python**
  - [x] add `trezorctl btc get-descriptor` command
  - [x] add tests for `trezorctl btc get-descriptor` command`
  - [x] update `get-descriptor` test to include correct `root_fingerprint`
* **legacy**
  - [x] implement `ignore_xpub_magic`
  - [x] implement `root_fingerprint`
  - [x] show XPUB instead of Pubkey in GetPublicKey screen
  - [x] add `ignore_xpub_magic` tests
  - [x] add `root_fingerprint` tests
* **core**
  - [x] implement `ignore_xpub_magic`
  - [x] implement `root_fingerprint`
  - [x] show XPUB instead of Pubkey in GetPublicKey screen
  - [x] add `ignore_xpub_magic` tests
  - [x] add `root_fingerprint` tests

### CODE REVIEW

**Comment by matejcik** [comment]
what is the reason for `ignore_xpub_magic`? wouldn't it be better for the client to strip/recalculate that?

**Comment by prusnak** [comment]
> what is the reason for `ignore_xpub_magic`? wouldn't it be better for the client to strip/recalculate that?

We need to show the same xpub on the display

**Comment by matejcik** [comment]
> We need to show the same xpub on the display

I'm wondering whether there should be a `GetDescriptor` call then. Comparing display XPUB and host-side full descriptor is weird.

OTOH I can see uses for the xpub display other than the descriptor, so :woman_shrugging:

File: python/src/trezorlib/btc.py [hunk 1]
```diff
@@ -74,10 +75,57 @@ def get_public_node(
             show_display=show_display,
             coin_name=coin_name,
             script_type=script_type,
+            ignore_xpub_magic=ignore_xpub_magic,
         )
     )
 
 
+def get_descriptor(
```
Comment (matejcik) [inline python/src/trezorlib/btc.py:83]: this function should move to `trezorlib.cli.btc` as it's too high-level for this file

it should also probably accept `address_n`, and the `get-descriptor` command should also accept `-n/--address`, plus possibly `-a/--account` as a convenience shortcut

File: python/src/trezorlib/btc.py [hunk 2]
```diff
@@ -74,10 +75,57 @@ def get_public_node(
...
+    elif script_type == messages.InputScriptType.SPENDP2SHWITNESS:
+        acc_type = 49
+        fmt = "sh(wpkh({}))"
+    elif script_type == messages.InputScriptType.SPENDWITNESS:
+        acc_type = 84
+        fmt = "wpkh({})"
+    else:
+        raise exceptions.TrezorException("Unsupported account type")
+
+    # TODO: correctly lookup SLIP44 for coin provided
```
Comment (matejcik) [inline python/src/trezorlib/btc.py:102]: In context of accepting `address_n`, it's not important that we don't have the full SLIP44 lookup here. (fwiw we don't bundle the data for this anymore)

**Comment by matejcik** [comment]
left some comments on python part

LGTM on core side

did not review legacy, i might look at it later if nobody else does

File: python/src/trezorlib/btc.py (hunk 1 above)
Comment (prusnak) [reply python/src/trezorlib/btc.py:83]: I prefer this won't accept  `address_n` - I don't see a good reason why it should. Will move this to `trezorlib.cli.btc`

Comment (prusnak) [reply python/src/trezorlib/btc.py:83]: If I moved this to `trezorlib.cli.btc`, where should I move unit test from 9af06e7de6b48f85efd32d9cd8a49d7276bc28da?

I really want to have this feature properly covered by tests.

Comment (matejcik) [reply python/src/trezorlib/btc.py:83]: well the reason to accept `address_n` is to allow use by coins other than Bitcoin, but i don't care that much so eh

re tests, we can keep them where they are, importing from `trezorlib.cli` namespace is not discouraged or anything

i'm thinking that the bulk of the functionality should be tested in `python/tests` presumably by mocking out the `btc.get_public_node` call?
(i don't think it is super important though)

**Comment by matejcik** [comment]
reviewed legacy as well, looks good

File: python/src/trezorlib/btc.py [hunk 3]
```diff
@@ -74,10 +75,57 @@ def get_public_node(
...
+        n,
+        ecdsa_curve_name=None,
+        show_display=show_display,
+        coin_name=coin_name,
+        script_type=script_type,
+        ignore_xpub_magic=True,
+    )
+
+    fingerprint = pub.root_fingerprint if pub.root_fingerprint is not None else 0
+    external = "[{:08x}{}]{}/0/*".format(fingerprint, path[1:], pub.xpub)
```
[the remark this answers is not in the document]
Comment (prusnak) [reply python/src/trezorlib/btc.py]: Addressed in 15c37531416c35656058c67c7f99345511c1f61c

File: python/src/trezorlib/btc.py [hunk 4]
```diff
@@ -74,10 +75,57 @@ def get_public_node(
...
+    internal = "[{:08x}{}]{}/1/*".format(fingerprint, path[1:], pub.xpub)
+
+    return (fmt.format(external), fmt.format(internal))
```
[the remark this answers is not in the document]
Comment (prusnak) [reply python/src/trezorlib/btc.py]: Addressed in 15c3753

File: python/src/trezorlib/btc.py (hunk 1 above)
Comment (prusnak) [reply python/src/trezorlib/btc.py]: Function moved to cli.btc in 15c3753

Tests moved in ec4d3c36aca2898a2ab1571f0766334c0fa13b9c

File: python/src/trezorlib/btc.py (hunk 2 above)
Comment (prusnak) [reply python/src/trezorlib/btc.py]: Let's drop the SLIP44 lookup completely and accept only Bitcoin/Testnet in the func

**Comment by prusnak** [comment]
I updated the tests to test the cli module in ec4d3c3, but now they fail with:

```
E           TypeError: 'TrezorClientDebugLink' object is not iterable
```

Can you please advise?

**Comment by prusnak** [comment]
> I updated the tests to test the cli module in [ec4d3c3](https://github.com/trezor/trezor-firmware/commit/ec4d3c36aca2898a2ab1571f0766334c0fa13b9c), but now they fail with:

Fixed in 1bad419561725e8328bbd6fb924abd9014a792a8

**Comment by prusnak** [comment]
This is now ready for review. One things I am unsure about:

There are two ways how to obtain the root fingerprint:
1. hack the derive function to return not only a derived node, but also a root fingerprint
2. derive a dummy node (e.g. `m/0'`) and return a parent fingerprint stored in that node

I used method 1 in Legacy and method 2 in Core. My reasoning is that using method 2 in Core would introduce lots of code/typing changes, but I do agree that method 2 is quite a hack. We might consider using method 2 for Legacy as well.

Let me know what you think.

**Comment by andrewkozlik** [comment]
> There are two ways how to obtain the root fingerprint:
> 
> 1. hack the derive function to return not only a derived node, but also a root fingerprint
> 2. derive a dummy node (e.g. `m/0'`) and return a parent fingerprint stored in that node
> 
> I used method 1 in Legacy and method 2 in Core. My reasoning is that using method 2 in Core would introduce lots of code/typing changes, but I do agree that method 2 is quite a hack. We might consider using method 2 for Legacy as well.
> 
> Let me know what you think.

Method 2 is cleaner, but I don't have a problem with either of the two approaches. It's no big deal if this is implemented slightly differently in core and legacy, so I'd say it's OK as is.

**Review by matejcik — APPROVED** [review]

**Comment by matejcik** [comment]
> There are two ways how to obtain the root fingerprint:
> 
> 1. hack the derive function to return not only a derived node, but also a root fingerprint
> 2. derive a dummy node (e.g. `m/0'`) and return a parent fingerprint stored in that node
> 
> I used method 1 in Legacy and method 2 in Core. My reasoning is that using method 2 in Core would introduce lots of code/typing changes, but I do agree that method 2 is quite a hack. We might consider using method 2 for Legacy as well.

I'm a little confused here, ISTM both core and legacy are using the same method?
Personally I'd think the cleanest solution would be a `root_fingerprint()` function in crypto, but that seems needless, the employed solution looks good enough

**Comment by prusnak** [comment]
> I'm a little confused here, ISTM both core and legacy are using the same method?

I reworked it to use the same method and force-pushed.

**Comment by prusnak** [comment]
Can you please @andrewkozlik have one last look at reworked legacy change (that uses the same method as core, i.e. deriving m/0'): https://github.com/trezor/trezor-firmware/pull/1363/commits/c76145afdc6da826a447f316189f955e98714c3d

**Review by andrewkozlik — APPROVED** [review]
utACK for the root_fingerprint derivation.
However, note that the UI test are failing.

**Comment by prusnak** [comment]
Hm, about the tests: for some strange reason I'm getting different hashes each time I run `pytest test_descriptors.py --ui=record`. Can someone else have a look at it? Maybe @tsusanka has an idea?

**Comment by andrewkozlik** [comment]
> Hm, about the tests: for some strange reason I'm getting different hashes each time I run `pytest test_descriptors.py --ui=record`. Can someone else have a look at it? Maybe @tsusanka has an idea?

I suppose it could be the pulsating swipe arrow, but we have the same thing in test_msg_backup_device.py, so the UI tests should be accustomed to it.

**Comment by prusnak** [comment]
Ah, it's probably the same reason why `test_msg_getpublickey.py` has `pytestmark = pytest.mark.skip_ui`. Can you please confirm @matejcik that's the reason?

**Comment by tsusanka** [comment]
> I suppose it could be the pulsating swipe arrow, but we have the same thing in test_msg_backup_device.py, so the UI tests should be accustomed to it.

Yes, the pulsing should disabled as the tests are run against emulator with disabled animations.

**Comment by prusnak** [comment]
I confirm I have reproducible hashes using `TREZOR_DISABLE_ANIMATION=1` when running the tests. 

Let's wait for the CI.

**Comment by tsusanka** [comment]
Yes and it is in the [Makefile](https://github.com/trezor/trezor-firmware/blob/fe55f0c1f0e7fff1fbd75a75cfd84e9bda7c9fdc/core/Makefile#L51) so we should be all good.

**Pull request merged**